Adding text editor functionality to text viewer.

nevumx opened the pull request:
Pursuant to my feature request https://github.com/d0k3/GodMode9/issues/861, I am pleased to present my proposal PR for adding text editor functionality to the existing text viewer by repurposing the existing keyboard that is used for file renames.

Features:
- UTF-8 display and typing support
- Support for both LF and CRLF line endings, not CR line endings
- PGUP/PGDN/HOME/END-style cursor functionality
- Full wordwrapped/non-wordwrapped text support

Screenshots:
![snap_240627192950](https://github.com/d0k3/GodMode9/assets/8940345/ec2f33b0-050f-42ef-a0b9-9a08c7e897bd)
![snap_240627161019](https://github.com/d0k3/GodMode9/assets/8940345/dedcd2a8-19d5-4bcc-8c33-e2f6e01fb03d)
![snap_240627161030](https://github.com/d0k3/GodMode9/assets/8940345/f96d655b-a138-4bf3-8324-9635c4727ec6)
![snap_240627161239](https://github.com/d0k3/GodMode9/assets/8940345/691baf43-ab66-4a84-b9ea-3dbdd1a3887e)
![snap_240627161245](https://github.com/d0k3/GodMode9/assets/8940345/389efbd0-a09d-4ec6-8531-cae3121569ec)

Considerations:
- Some functions shared by, for example, the README ToC Viewer and System information viewer has been changed.
  - I tested the system information viewer with no issues
  - The README viewer seems to have disappeared for some reason, so I couldn't test that... 🤔
- In order to fully support editing both wordwrapped and non-wordwrapped text, there were some edge cases that resulted in some unsightly `if` statements around, for example, when the last line is a certain length.
- I tried to get the cursor to blink first by using code like `InputWait(1)`, but that resulted in inputs getting dropped occasionally when pressing a button between `InputWait` calls when the screen was being drawn.
  - I also tried a multi-threaded approach, but even though the code I wrote with `<threads.h>` compiled, the thread would never spawn due to [`thrd_error`](https://en.cppreference.com/w/c/thread/thrd_errors)... I am guessing threads are out of scope for this project, since I don't see them anywhere else, definitely open to other suggestions on this...
- I renamed the "Textviewer" to "Text Editor," which might look kind of odd next to "Hexeditor;" but "Texteditor" seemed weird too, so I am open to suggestions here too...
- Some hardware buttons used on the rename prompt as keyboard hotkeys are different than in the text editor.
  - For example, "R" doesn't switch between capslock states, instead it is used to move the cursor further, like with the text viewer.
  - Instead "Y" switches between capslock states, because I don't see a purpose for a hardware button to insert a space into the text...
  - "A" Adds a newline char instead of submitting.

Definitely open to sugestions on this, especially as it relates to key bindings/text/etc.

nevumx commented on README.md:
```diff
@@ -116,7 +116,7 @@ With the possibilites GodMode9 provides, not everything may be obvious at first
 * __Search drives and folders__: Just press R+A on the drive / folder you want to search.
 * __Compare and verify files__: Press the A button on the first file, select `Calculate SHA-256`. Do the same for the second file. If the two files are identical, you will get a message about them being identical. On the SDCARD drive (`0:`) you can also write an SHA file, so you can check for any modifications at a later point.
 * __Hexview and hexedit any file__: Press the A button on a file and select `Show in Hexeditor`. A button again enables edit mode, hold the A button and press arrow buttons to edit bytes. You will get an additional confirmation prompt to take over changes. Take note that for certain files, write permissions can't be enabled.
-* __View text files in a text viewer__: Press the A button on a file and select `Show in Textviewer` (only shows up for actual text files). You can enable wordwrapped mode via R+Y, and navigate around the file via R+X and the dpad.
```
Open to "Text Editor"/"Texteditor"/"Hex Editor"/"Hexeditor" guidance here...

nevumx commented on arm9/source/common/swkbd.h:
```diff
@@ -53,9 +65,9 @@ enum {
     KEY_ALPHA, ' ', KEY_BKSPC
 
 #define SWKBD_KEYS_NUMPAD \
-    '7', '8', '9', 'F', 'E', \
```
I changed the arrangement of these to be more consistent with e.g. calculators.

nevumx commented on arm9/source/common/swkbd.h:
```diff
@@ -68,12 +80,20 @@ enum {
     6, 0, \
     6, 0, \
     6, 0, \
-    3, 32, 46, 32, 0, \
```
This was bothering me as well.

nevumx commented on arm9/source/common/swkbd.h:
```diff
@@ -87,3 +107,5 @@ enum {
 
 #define ShowKeyboardOrPrompt (TouchIsCalibrated() ? ShowKeyboard : ShowStringPrompt)
 bool PRINTF_ARGS(3) ShowKeyboard(char* inputstr, u32 max_size, const char *format, ...);
+bool BuildKeyboard(TouchBox* swkbd, const char* keys, const u8* layout, bool multi_line);
```
This is exposed to prevent rebuilds between keypresses in the `MemTextViewer` function.

BlueRaja replied:
A lot of these Github comments should probably be code comments instead

nevumx replied:
@BlueRaja done. I went through and added code comments where I felt they were relevant. If you have any other concerns, please let me know. 😁

nevumx commented on arm9/source/godmode.c:
```diff
@@ -1225,7 +1225,7 @@ u32 FileHandlerMenu(char* current_path, u32* cursor, u32* scroll, PaneData** pan
     int n_opt = 0;
     int special = (special_opt) ? ++n_opt : -1;
     int hexviewer = ++n_opt;
-    int textviewer = (filetype & TXT_GENERIC) ? ++n_opt : -1;
```
Allowing for editing of empty files.

nevumx commented on arm9/source/godmode.c:
```diff
@@ -1316,6 +1316,7 @@ u32 FileHandlerMenu(char* current_path, u32* cursor, u32* scroll, PaneData** pan
     }
     else if (user_select == textviewer) { // -> show in text viewer
         FileTextViewer(file_path, scriptable);
+        GetDirContents(current_dir, current_path);
```
This is necessary, I assume, to recalculate file sizes?

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -259,86 +265,125 @@ static inline u32 hexntostr(const u8* hex, char* str, u32 len) {
 
-static inline u32 line_len(const char* text, u32 len, u32 ww, const char* line, char** eol) {
-    u32 last = ((text + len) - line);
+static inline bool is_crlf(const char* str) {
```
Majority rule seems to make the most sense here... We _could_ do a hotkey to switch newline types on the keyboard, and display the current newline type status on the keyboard screen too, let me know if you think that's a good idea.

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -259,86 +265,125 @@ static inline u32 hexntostr(const u8* hex, char* str, u32 len) {
+    return chr[0] == '\n' || (chr[0] == '\r' && chr[1] == '\n');
+}
+
+static inline u32 bytes_in_chars_u32(const char* str, u32 nchars) {
```
These are for slightly saner multibyte character support.

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -259,86 +265,125 @@ static inline u32 hexntostr(const u8* hex, char* str, u32 len) {
+    return chars;
+}
+
+static inline u32 line_len_chars(const char* text, u32 len, u32 ww, const char* line, const char** eol) {
```
I made these a bit more simplified while supporting multibyte chars, but as a result, their behavior has changed slightly. I changed invocations accordingly where necessary.

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -259,86 +265,125 @@ static inline u32 hexntostr(const u8* hex, char* str, u32 len) {
     }
 }
 
+static inline const char* line_start(const char* text, u32 len, u32 ww, const char* ptr) {
```
For example, this can be used now instead of `line_seek(..., 0);`

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -1614,7 +1659,7 @@ bool run_line(const char* line_start, const char* line_end, u32* flags, char* er
 
 // checks for illegal ASCII symbols
 bool ValidateText(const char* text, u32 len) {
-    if (!len) return false;
```
Allow editing empty files.

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
     // display text on screen
-    char txtstr[TV_LLEN_DISP + 1];
-    char* ptr = line0;
+    char txtstr[TV_LLEN_DISP * MAX_CHAR_SIZE + 1];
```
Add UTF-8 char max size.

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
-        if (ncpy > TV_LLEN_DISP) ncpy = TV_LLEN_DISP;
-        bool al = !ww && off_disp && (ptr != ptr_next);
-        bool ar = !ww && (llen > off_disp + TV_LLEN_DISP);
+        int off_disp_bytes = bytes_in_chars_int(ptr, off_disp_chars);
```
Multibyte support.

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
         if (TV_LNOS > 0) { // line number
-            if (ptr != ptr_next)
-                DrawStringF(TOP_SCREEN, x_lno, y, ((ptr == text) || (*(ptr-1) == '\n')) ? COLOR_TVOFFS : COLOR_TVOFFSL, COLOR_STD_BG, "%0*lu", TV_LNOS, nln);
+            bool prev_ww_line_full = ww && ww == chars_between_pointers(line_seek_chars(text, len, ww, ptr, -1), ptr);
```
Unfortunate edge case to draw line number when the `'\0'` is word-wrapped onto its own line...

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
+            u32 cursor_line_offset_chars = chars_between_pointers(ptr + off_disp_bytes, cursor);
+            if (cursor >= ptr + off_disp_bytes && cursor <= ptr + off_disp_bytes + ncpy_bytes && cursor_line_offset_chars < TV_LLEN_DISP
+                    && (cursor != ptr + off_disp_bytes + ncpy_bytes || is_newline(cursor) || cursor == text + len)) {
+                DrawRectangle(TOP_SCREEN, x_txt + cursor_line_offset_chars * FONT_WIDTH_EXT, y, FONT_WIDTH_EXT, 1, COLOR_RED);
```
Without a blinking line, a hollow red rect is the next best cursor I could think of. These coords are picked so that they are drawn over by text.

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
+            if (cursor >= ptr + off_disp_bytes && cursor <= ptr + off_disp_bytes + ncpy_bytes && cursor_line_offset_chars < TV_LLEN_DISP
+                    && (cursor != ptr + off_disp_bytes + ncpy_bytes || is_newline(cursor) || cursor == text + len)) {
+                DrawRectangle(TOP_SCREEN, x_txt + cursor_line_offset_chars * FONT_WIDTH_EXT, y, FONT_WIDTH_EXT, 1, COLOR_RED);
+                DrawRectangle(TOP_SCREEN, x_txt + (cursor_line_offset_chars + 1) * FONT_WIDTH_EXT - ((cursor_line_offset_chars == TV_LLEN_DISP - 1) ? 1 : 0), y, 1, FONT_HEIGHT_EXT, COLOR_RED);
```
Small edge case to squish the cursor box at the end of the line/screen.

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -1710,56 +1813,181 @@ bool MemTextViewer(const char* text, u32 len, u32 start, bool as_script) {
+            while (cursor && line0_next < line_seek_chars(text, len, ww, GetNextChar(cursor), -TV_NLIN_DISP)) line0_next = line_seek_chars(text, len, ww, line0_next, 1);
+        }
+
+        // find last allowed lines (ww and nonww)
```
These checks have to be performed every iteration now.

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -1710,56 +1813,181 @@ bool MemTextViewer(const char* text, u32 len, u32 start, bool as_script) {
+        if (save_path) {
+            bool diffs = false;
+            if (len != text_cpy_len) diffs = true;
+            else for (u32 i = 0; i < len; ++i) if (text[i] != text_cpy[i]) { diffs = true; break; }
```
This could be slow for large files, but I took it form the hex editor; open to suggestions here...

nevumx commented:
Not exactly sure who to tag here... @Wolfvak / @d0k3 ? Anyone else?

nevumx commented on arm9/source/utils/scripting.c:
```diff
@@ -1846,18 +2074,20 @@ bool MemToCViewer(const char* text, u32 len, const char* title) {
     // text file needs to fit inside the STD_BUFFER_SIZE
-    u32 flen, len;
+    size_t fileSize = FileGetSize(path);
+    if (fileSize >= STD_BUFFER_SIZE) {
```
Max file size check. I tested with a much lower number as well to make sure that the editor would not allow characters to be typed beyond the max file size.

Trickiy commented:
Could the hex editing functionality be added to the hex viewer as well?

nevumx commented:
> Could the hex editing functionality be added to the hex viewer as well?

The hex viewer has had edit functionality since 2016: https://github.com/d0k3/GodMode9/commit/88a62d8f2ecc50f1ee99e0a08359ae85d02dd0f3

nevumx commented:
FYI @d0k3 / @Wolfvak I have added clipboard select/cut/copy/paste support as well. The description has been updated!

nevumx commented:
@d0k3 / @Wolfvak, any more thoughts on this PR? I think text editing would be a nice feature to have, for example, for editing config files directly in GM9 without the need for more software.

nevumx commented:
@d0k3 / @Wolfvak @BlueRaja, any more thoughts on this PR? I think text editing would be a nice feature to have, for example, for editing config files directly in GM9 without the need for more software.

nevumx commented:
@d0k3 / @Wolfvak / @BlueRaja Bumping this

d0k3 commented:
@nevumx - sorry about the long radio silence, know that your contribution is appreciated. As you see, I didn't merge this p/r for the current release, however that doesn't mean it won't be in a future release. Could you rebase this on the current master branch?

nevumx commented on arm9/source/lua/gm9ui.c:
```diff
@@ -206,12 +206,16 @@ static int ui_show_text_viewer(lua_State* L) {
     // validate text ourselves so we can return a better error
     // MemTextViewer calls ShowPrompt if it's bad, and i don't want that
+
+    if (len >= STD_BUFFER_SIZE) {
```
Adding this to be more in line with `FileTextViewer(...)` code.

nevumx commented on arm9/source/lua/gm9ui.c:
```diff
@@ -225,6 +229,11 @@ static int ui_show_file_text_viewer(lua_State* L) {
     // MemTextViewer calls ShowPrompt if it's bad, and i don't want that
     // and FileTextViewer calls the above function
+
+    size_t fileSize = FileGetSize(path);
```
Adding this to be more in line with `FileTextViewer(...)` code.

nevumx commented on arm9/source/lua/gm9ui.c:
```diff
@@ -233,16 +242,14 @@ static int ui_show_file_text_viewer(lua_State* L) {
     size_t flen = FileGetData(path, text, STD_BUFFER_SIZE - 1, 0);
-
     text[flen] = '\0';
-    u32 len = (ptrdiff_t)memchr(text, '\0', flen + 1) - (ptrdiff_t)text;
```
This line is unnecessary; using `flen` instead as we do in the `FileTextViewer(...)` code.

nevumx commented:
> @nevumx - sorry about the long radio silence, know that your contribution is appreciated. As you see, I didn't merge this p/r for the current release, however that doesn't mean it won't be in a future release. Could you rebase this on the current master branch?

@d0k3 No problem, I rebased, and fixed the Lua invocations of `MemTextViewer(...)` to be more in line with the `FileTextViewer(...)` code. I did a quick smoke test on my 3DS to make sure everything was working well, which it was. I don't know how to test Lua though, so I would just give that a quick spin before merging. That being said, I can't imagine anything being much different with the `MemTextViewer(...)` calls I see in the Lua code.

nevumx commented:
@d0k3 How is the outlook for getting this merged? 😅
